### PR TITLE
docs: Add example for DARRAY as function parameter

### PR DESCRIPTION
**Pull request by CodeYan01**
### Description

Add example for DARRAY as function parameter.
image
image

While I'm at it, I also fixed the DARRAY macro definition in the docs (which previously did not show up because of a wrong directive).

### Motivation and Context

When I first started using DARRAY, it was very confusing, and there isn't any real documentation on how to use them other than code in the obs-studio repo.

### How Has This Been Tested?

Built and viewed locally. As for the code example, I tried that myself and it worked correctly, and the integers in the array were incremented.

### Types of changes

- Documentation (a change to documentation pages)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by CodeYan01** [comment]
Just gonna retest. I'm not really sure why functions have to set the function parameter back to the .da of the dynamic array in the function at the end of the function, and removing it seems to cause stuff to not work, but in my example, i'm pretty sure the integers were edited.

**Comment by norihiro** [comment]
> I'm not really sure why functions have to set the function parameter back to the .da of the dynamic array in the function at the end of the function

If you are talking a line like below, this is because the function `insert_packet` modified the dynamic array and the pointer in the structure could be reallocated. If the reallocation happens, the contents of the structure has to be copied back to the original structure.
https://github.com/obsproject/obs-studio/blob/a7dffb98a5dec3721b39612cad776dd13582426b/plugins/obs-ffmpeg/obs-ffmpeg-mux.c#L1154

**Comment by CodeYan01** [comment]
@norihiro Thank you! I finally understand why as you've explained it clearly. I didn't know the array had to be realloc'd (but it makes sense, as otherwise we can't access it with array indexing). Will adjust the example to show this.

**Comment by CodeYan01** [comment]
I made no mention of reallocation, because i think, even if the array isn't realloc'd (because sometimes its capacity is bigger than num, so adding items wouldn't resize it), darray size changes still need to be synced back to the original one, since the num in the original wouldn't be updated otherwise

**Review by Lain-B — COMMENTED** [review]
I actually disagree with suggesting to programmers the `da` trick. The reason why is because it's very unsafe. I shouldn't even have done it in my prior code.

Instead, I would recommend a note like this:
```
**IMPORTANT NOTE:** If you need to pass dynamic arrays as a parameter to
a function, it's generally recommended to keep them as a member variable
of a structure and then pass a pointer to that structure, even if it's
the only variable within the structure. While you can access the
internal **darray** structure via the **da** member variable, using it
is much more unsafe, so if you must pass a singular dynamic array to a
function, it's recommended to just wrap it in a structure and pass that
structure instead.
```

What do you all think?

**Comment by CodeYan01** [comment]
I agree that the da trick is unsafe, as I've seen first-hand multiple times, but I think I like @norihiro's approach better https://github.com/obsproject/obs-studio/pull/9072 where they simply added a define for the darray at the top of the file, which is used for every function argument. that way, you don't need the da trick, and you pass the darray directly as a parameter. I think that putting the darray in a struct instead is another level of indirection, which means longer code. What do you think about @norihiro's approach?

**Comment by Lain-B** [comment]
I think @norihiro's method is indeed safer, I'm more comfortable with that. You could probably change my commentary to something like "if you need to pass it as a parameter, use it in a structure or create a typedef"

**Comment by CodeYan01** [comment]
Updated. Tried to integrate the note suggested by @Lain-B, though i had to rephrase it in consideration of the example. I picked the typedef to write an example for, as that would make for a shorter example.

File: docs/sphinx/reference-libobs-util-darray.rst [hunk 1]
```diff
@@ -50,6 +50,42 @@ dynamic array value with a reference (&) operator.  For example:
...
+the dynamic array directly. An example with the typedef method:
+
+.. code:: cpp
+
+   typedef DARRAY(int) int_array_t;
+
+   void generate_integers(int_array_t *integers, int start, int end)
+   {
+           for (int i = start; i < end; i++)
+                   da_push_back(*integers, &i);
```
Comment (CodeYan01) [inline docs/sphinx/reference-libobs-util-darray.rst:66]: For this, I removed the parentheses that is necessary atm, like `da_push_back((*integers), &i);`, due to conversation at https://github.com/obsproject/obs-studio/pull/9072#discussion_r1236385202

**Comment by CodeYan01** [comment]
I added back the parentheses around the darray in `da_push_back((*integers), &i);` despite the convo at https://github.com/obsproject/obs-studio/pull/9072#discussion_r1236385202, as I realized that most plugins still build against older versions.

**Comment by CodeYan01** [comment]
If norihiro's PR adding parentheses in the macro gets merged, I could then update the docs and add `versionchanged::`

**Comment by Lain-B** [comment]
Merge @norihiro's changes. Feel free to update.

**Comment by Lain-B** [comment]
I'm just going to add the modification and merge it manually. You'll see "closed in [commit]"

**Comment by CodeYan01** [comment]
I apologize, I didn't have time this week to look at it. I'll just write a draft PR for the versionchanged directive

**Pull request merged**